crates/bevy_pbr/src/render/light.rs

```diff
@@ -1753,17 +1617,34 @@ pub fn queue_shadows(
             // NOTE: Lights with shadow mapping disabled will have no visible entities
             // so no meshes will be queued
             for entity in visible_entities.iter().copied() {
-                if let Ok(mesh_handle) = casting_meshes.get(entity) {
-                    if let Some(mesh) = render_meshes.get(mesh_handle) {
-                        let mut key =
-                            ShadowPipelineKey::from_primitive_topology(mesh.primitive_topology);
+                if let Ok((mesh_handle, material_handle)) = casting_meshes.get(entity) {
+                    if let (Some(mesh), Some(material)) = (
+                        render_meshes.get(mesh_handle),
+                        render_materials.get(material_handle),
+                    ) {
+                        let mut mesh_key =
+                            MeshPipelineKey::from_primitive_topology(mesh.primitive_topology)
+                                | MeshPipelineKey::DEPTH_PREPASS;
                         if is_directional_light {
-                            key |= ShadowPipelineKey::DEPTH_CLAMP_ORTHO;
+                            mesh_key |= MeshPipelineKey::DEPTH_CLAMP_ORTHO;
+                        }
+                        let alpha_mode = material.properties.alpha_mode;
+                        match alpha_mode {
+                            AlphaMode::Mask(_) => {
+                                mesh_key |= MeshPipelineKey::ALPHA_MASK;
+                            }
+                            AlphaMode::Blend | AlphaMode::Add => {
```

Comment:
The reason I did not include `AlphaMode::Premultiplied` was that e.g. `(1,1,1,0)` would produce a visible fragment, so we cannot use the alpha value alone as the discard criterion. But we could probably use the maximum component instead. And regarding `Add` I would prefer the opt-out via `NotShadowCaster` for more consistency, but I am open to disabling them by default.

Reply:
Yeah, I'd prefer opt out too, it's always nice to have control on these things compared to making it implicit.

Reply:
I've now included the `AlphaMode::Premultiplied` case too. The fragments are discarded if all component values are smaller than the `PREMULTIPLIED_ALPHA_CUTOFF` constant. This matches the behavior of `AlphaMode::Blend` when the RGB values really were premultiplied with an alpha value smaller than `PREMULTIPLIED_ALPHA_CUTOFF`, because then they cannot be larger than `PREMULTIPLIED_ALPHA_CUTOFF`.